src/server/models/url.ts

```diff
@@ -0,0 +1,163 @@
+import Sequelize from 'sequelize'
+import { parse } from 'url'
+import { ACTIVE, INACTIVE } from './types'
+import blacklist from '../resources/blacklist'
+import { isHttps } from '../../shared/util/validation'
+import { sequelize } from '../util/sequelize'
+import { IdType } from '../../types/server/models'
+import { ogHostname } from '../config'
+
+interface UrlBaseType extends IdType {
+    readonly shortUrl: string
+    readonly longUrl: string
+    readonly state: Sequelize.EnumDataType<string>
+    readonly isFile: boolean
+}
+
+
+export interface UrlType extends IdType, UrlBaseType, Sequelize.Model {
+    readonly clicks: Sequelize.IntegerDataType
+}
+
+// For sequelize define
+type UrlTypeStatic = typeof Sequelize.Model & {
+    new(values?: object, options?: Sequelize.BuildOptions): UrlType
+}
+
+export const Url = <UrlTypeStatic>sequelize.define('url', {
+  shortUrl: {
+    type: Sequelize.STRING,
+    primaryKey: true,
+    validate: {
+      is: /^[a-z0-9-]+$/,
+    },
+  },
+  longUrl: {
+    type: Sequelize.TEXT, // Support >255 chars
+    validate: {
+      isUrl: true,
+
+      httpsCheck(url: string) { // Must be https
+        if (!isHttps(url)) throw new Error('Only HTTPS URLs are allowed.')
+      },
+
+      noCircularRedirects(url: string) {
+        if (parse(url).hostname === ogHostname) {
+          throw new Error('Circular redirects to go.gov.sg are prohibited')
+        }
+      },
+
+      // Blacklist check
+      blacklistCheck(longUrl: string) {
+        if (blacklist.some((bl) => longUrl.includes(bl))) {
+          throw new Error(
+            'Database creation of URLs to link shortener sites prohibited.',
+          )
+        }
+      },
+    },
+    allowNull: false,
+  },
+  state: {
+    type: Sequelize.ENUM,
+    values: [ACTIVE, INACTIVE],
+    defaultValue: ACTIVE,
+  },
+  clicks: {
+    type: Sequelize.INTEGER,
+    allowNull: false,
+    defaultValue: 0,
+  },
+  isFile: {
+    type: Sequelize.BOOLEAN,
+    allowNull: true, // TODO: Make this false after backfill.
+  },
+}, {
+  hooks: {
+    afterCreate: async (url: UrlType, options) => {
+      if (!options.transaction) {
+        return Promise.reject(new Error('URL creation must be wrapped in a transaction'))
+      }
+      // eslint-disable-next-line no-use-before-define
+      await writeToUrlHistory(
+        url,
+                options as Sequelize.CreateOptions & { transaction: Sequelize.Transaction },
```

Comment:
indentation looks a bit broken in many places like this. did any lint errors pop up for you? if not then something is fishy about our lint config

Reply:
There aren't any errors. Seems to have happened with VSCode formatting using our lint config, will check if it's my settings. I'll manually fix these first.